Implement macro meta-variable expressions

c410-f3r opened the pull request:
Based on https://github.com/markbt/rust/tree/metavariable_expressions with some refactoring, additional tests, explicit feature declaration and minor documentation enhancement.

With this feature, declarative macros will be much more powerful and more flexible to work with.

```rust
#![feature(macro_metavar_expr)]

macro_rules! build_and_modify_tuple {
    ( ( $( $l:literal ),* ) ) => {{
        let mut tuple = ( $( $l ),* );
        $( ${ignore(l)} tuple.${index()} *= 2; )*
        tuple
    }};
}

macro_rules! count {
    ( $( $e:stmt ),* ) => {
        ${ count(e) }
    };
}

fn main() {
    assert_eq!(build_and_modify_tuple!((0u8, 2i32, 4u128)), (0u8, 4i32, 8u128));
    assert_eq!(count!(let _a = 1i32, let _ = (), if true {}), 3);
}
```

Thank you very much @markbt for writing the RFC and for starting the implementation.

cc https://github.com/rust-lang/rust/issues/83527
r? @petrochenkov

c410-f3r commented on src/test/ui/rfc-3086-metavar-expr/nested-macro-declarations.rs:
```diff
@@ -0,0 +1,28 @@
+            ( $$( $b:ident ),* ) => {
+                $$(
+                    macro_rules! $b {
+                        ( $$$$( $c:ident ),* ) => {
+                            $$$$(
+                                fn $c() -> &'static str { stringify!($c) }
+                            ),*
+                        };
```
The more nested, the more double dollar signs will be necessary. Not sure if it is correct

c410-f3r replied:
https://github.com/rust-lang/rfcs/pull/3086#discussion_r589286116

c410-f3r commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -352,6 +357,38 @@ impl LockstepIterSize {
+    }
+    for &(idx, _) in repeats {
+        if let MatchedSeq(ref ads) = matched {
+            if let Some(elem) = ads.get(idx) {
+                matched = elem;
+            }
```
Not found indices are ignored

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,285 @@
+        while let Some(tt) = tts.next() {
+            match tt {
+                tokenstream::TokenTree::Token(token) if token.is_ident() => {
+                    if let Some(ident) = token.ident() {
+                        result.push(Arg::Ident(ident.0))
+                    }
```
Also ignores if not found instead of `token.ident().unwrap()`. IIRC, this uncovered an error that was being silenced

c410-f3r commented on compiler/rustc_expand/src/mbe/macro_check.rs:
```diff
@@ -278,6 +278,7 @@ fn check_binders(
                 binders.insert(name, BinderInfo { span, ops: ops.into() });
             }
         }
+        TokenTree::MetaVarExpr(..) => {}
```
I am also not sure if this branch needs to be evaluated.

c410-f3r reviewed (commented):
If more documentation needs to be created, then I will do so in a following PR.

fmease commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -385,6 +422,94 @@ fn lockstep_iter_size(
+                    None => {
+                        return Err(cx.struct_span_err(
+                            sp.entire(),
+                            &format!("variable '{}' does not repeat at this depth", ident),
```
```suggestion
                            &format!("variable `{}` does not repeat at this depth", ident),
```

fmease commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -385,6 +422,94 @@ fn lockstep_iter_size(
+                    return Err(cx.struct_span_err(
+                        sp.entire(),
+                        &format!(
+                            "variable '{}' is not recognised in meta-variable expression",
```
```suggestion
                            "variable `{}` is not recognised in meta-variable expression",
```

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5022764109?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/4f2ec055c5c5567ceace0c0e30720d127ce983e4/checks/5022764109/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] ui/parser/macro/macro-repeat.rs stdout ----
diff of stderr:

- error: variable 'v' is still repeating at this depth
+ error: variable `v` is still repeating at this depth
3    |
4 LL |         $v

5    |         ^^
5    |         ^^
6 
- error: variable 'v' is still repeating at this depth
+ error: variable `v` is still repeating at this depth
9    |
10 LL |         $v



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/macro/macro-repeat/macro-repeat.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args parser/macro/macro-repeat.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/parser/macro/macro-repeat.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/macro/macro-repeat" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/parser/macro/macro-repeat/auxiliary"
------------------------------------------

------------------------------------------
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
stderr:
------------------------------------------
error: variable `v` is still repeating at this depth
   |
LL |         $v
   |         ^^


error: variable `v` is still repeating at this depth
   |
LL |         $v
   |         ^^


error: aborting due to 2 previous errors


------------------------------------------


---- [ui] ui/rfc-3086-metavar-expr/syntax-errors.rs stdout ----


78 LL |     ( $i:ident ) => { ${ count($i) } };
80 
- error: variable 'i' is still repeating at this depth
- error: variable 'i' is still repeating at this depth
+ error: variable `i` is still repeating at this depth
83    |
83    |
84 LL |     ( $( $i:ident ),* ) => { count($i };

The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-3086-metavar-expr/syntax-errors/syntax-errors.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args rfc-3086-metavar-expr/syntax-errors.rs`
error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-3086-metavar-expr/syntax-errors" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-3086-metavar-expr/syntax-errors/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: mismatched closing delimiter: `}`
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:47:35
   |
LL |     ( $( $i:ident ),* ) => { count($i };
   |                            -      ^   ^ mismatched closing delimiter
   |                            |      unclosed delimiter
   |                            closing delimiter possibly meant for this

error: expected meta-variable expression argument
error: expected meta-variable expression argument
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:17:39
   |
LL |     ( $( $i:ident ),* ) => { ${ count($i) } };

error: expected ','
error: expected ','
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:17:40
   |
LL |     ( $( $i:ident ),* ) => { ${ count($i) } };

error: count requires an identifier to count
error: count requires an identifier to count
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:17:33
   |
LL |     ( $( $i:ident ),* ) => { ${ count($i) } };


error: expected one of: `*`, `+`, or `?`
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:17:31
   |
LL |     ( $( $i:ident ),* ) => { ${ count($i) } };

error: expected meta-variable expression argument
error: expected meta-variable expression argument
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:26:32
   |
LL |     ( $i:ident ) => { ${ count($i) } };

error: expected ','
error: expected ','
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:26:33
   |
LL |     ( $i:ident ) => { ${ count($i) } };

error: count requires an identifier to count
error: count requires an identifier to count
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:26:26
   |
LL |     ( $i:ident ) => { ${ count($i) } };


error: expected one of: `*`, `+`, or `?`
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:26:24
   |
LL |     ( $i:ident ) => { ${ count($i) } };


error: variable 'i' does not repeat at this depth
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:12:24
   |
LL |     ( $i:ident ) => { ${ count(i) } };


error: expected one of `.`, `;`, `?`, `}`, or an operator, found `count`
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:17:33
   |
LL |     ( $( $i:ident ),* ) => { ${ count($i) } };
   |                                 ^^^^^ expected one of `.`, `;`, `?`, `}`, or an operator
...
LL |     curly__rhs_dollar__round!(a, b, c);
   |
   |
   = note: this error originates in the macro `curly__rhs_dollar__round` (in Nightly builds, run with -Z macro-backtrace for more info)
error: attempted to repeat an expression containing no syntax variables matched as repeating at this depth
error: attempted to repeat an expression containing no syntax variables matched as repeating at this depth
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:26:24
   |
LL |     ( $i:ident ) => { ${ count($i) } };


error: variable `i` is still repeating at this depth
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:47:36
   |
LL |     ( $( $i:ident ),* ) => { count($i };

error[E0425]: cannot find function `count` in this scope
error[E0425]: cannot find function `count` in this scope
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:35:30
   |
LL |     ( $( $i:ident ),* ) => { count(i) };
...
...
LL |     no_curly__no_rhs_dollar__round!(a, b, c);
   |
   |
   = note: this error originates in the macro `no_curly__no_rhs_dollar__round` (in Nightly builds, run with -Z macro-backtrace for more info)
error[E0425]: cannot find value `i` in this scope
error[E0425]: cannot find value `i` in this scope
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:35:36
   |
LL |     ( $( $i:ident ),* ) => { count(i) };
...
...
LL |     no_curly__no_rhs_dollar__round!(a, b, c);
   |
   |
   = note: this error originates in the macro `no_curly__no_rhs_dollar__round` (in Nightly builds, run with -Z macro-backtrace for more info)
error[E0425]: cannot find function `count` in this scope
error[E0425]: cannot find function `count` in this scope
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:41:23
   |
LL |     ( $i:ident ) => { count(i) };
...
...
LL |     no_curly__no_rhs_dollar__no_round!(a);
   |
   |
   = note: this error originates in the macro `no_curly__no_rhs_dollar__no_round` (in Nightly builds, run with -Z macro-backtrace for more info)
error[E0425]: cannot find value `i` in this scope
error[E0425]: cannot find value `i` in this scope
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:41:29
   |
LL |     ( $i:ident ) => { count(i) };
...
...
LL |     no_curly__no_rhs_dollar__no_round!(a);
   |
   |
   = note: this error originates in the macro `no_curly__no_rhs_dollar__no_round` (in Nightly builds, run with -Z macro-backtrace for more info)
error[E0425]: cannot find function `count` in this scope
error[E0425]: cannot find function `count` in this scope
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:53:23
   |
LL |     ( $i:ident ) => { count($i) };
...
...
LL |     no_curly__rhs_dollar__no_round!(a);
   |
   |
   = note: this error originates in the macro `no_curly__rhs_dollar__no_round` (in Nightly builds, run with -Z macro-backtrace for more info)
error[E0425]: cannot find value `a` in this scope
error[E0425]: cannot find value `a` in this scope
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:65:37
   |
LL |     no_curly__rhs_dollar__no_round!(a);

error: aborting due to 19 previous errors

For more information about this error, try `rustc --explain E0425`.
For more information about this error, try `rustc --explain E0425`.

------------------------------------------



failures:
    [ui] ui/parser/macro/macro-repeat.rs
    [ui] ui/rfc-3086-metavar-expr/syntax-errors.rs
test result: FAILED. 12456 passed; 2 failed; 121 ignored; 0 measured; 0 filtered out; finished in 98.73s

Build completed unsuccessfully in 0:10:19
```

</details>

c410-f3r commented on src/test/ui/rfc-3086-metavar-expr/feature-gate-macro_metavar_expr.rs:
```diff
@@ -0,0 +1,139 @@
+    ( $( { $( [ $( $outer:ident : ( $( $inner:ident )* ) )* ] )* } )* ) => {
+        [
+            (
+                ${count(inner)},
+                ${count(inner, 0)},
+                ${count(inner, 1)},
+                ${count(inner, 2)},
+                ${count(inner, 3)},
+                ${count(inner, 4)},
+                ${count(inner, 5)},
```
"Out-of-bound" depths are saturated to the innermost depth count

c410-f3r replied:
This is a bit strange because https://github.com/markbt/rfcs/blob/macro_metavar_expr/text/0000-macro-metavar-expr.md#count explicitly says that count is referring outer-most values.

> The three values this expands to are the number of outer-most repetitions (the number of times a would be generated), the sum of the number of middle repetitions (the number of times b would be generated), and the total number of repetitions of $x.

While `index` and `length` are referring to inner-most values (https://github.com/markbt/rfcs/blob/macro_metavar_expr/text/0000-macro-metavar-expr.md#index-and-length).

> The three values this expands to are the number of outer-most repetitions (the number of times a would be generated), the sum of the number of middle repetitions (the number of times b would be generated), and the total number of repetitions of $x.

In any case, `count` now stars at `0` instead of `1`. The old behavior was indeed confusing.

joshtriplett commented:
Could you please separate out the change from `'{}'` to `` `{}` `` and the corresponding test changes, for separate review and FCP? That seems orthogonal to this change, and it'd reduce the amount that needs to be reviewed simultaneously.

petrochenkov commented on compiler/rustc_expand/src/mbe/quoted.rs:
```diff
@@ -197,7 +235,12 @@ fn parse_tree(
-                // `tree` is followed by a random token. This is an error.
+                // `tree` is followed by another `$`. This is an escaped `$`.
+                Some(tokenstream::TokenTree::Token(Token { kind: token::Dollar, span })) => {
+                    TokenTree::token(token::Dollar, span)
```
`$$` escaping is not feature-gated.

petrochenkov commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -385,6 +422,94 @@ fn lockstep_iter_size(
+            match interp.get(&ident) {
+                Some(matched) => match count_repetitions(matched, &repeats, depth) {
+                    Some(count) => {
+                        marker.visit_span(&mut sp.entire());
```
What does this line mean?
(And other `marker.visit_span(&mut sp.entire())` lines below.)

c410-f3r replied:
Heh... I am not sure... :)

They do give the span an additional context but it seems that this additional context is not necessary. Therefore, I removed all `marker.visit_span`.

petrochenkov commented:
I still don't understand two things:
- which of the PR implements `ignore`, I don't see where variables used inside `${...}` are marked as used, but apparently the variables used inside `ignore` are indeed counted as used because there's no warning.
- what is the difference between `length` and `count`.

petrochenkov commented:
We need tests making sure that all kinds of erroneous conditions produce errors:
- The variables named inside meta-variable expressions are defined.
- The indices used inside meta-variable expressions are in range, I don't think there should be any silent saturating or wrapping behavior.

petrochenkov commented:
Also, parsing functions returning `Option` are error-prone.
Cases in which the syntactic element is truly optional, and in which it's expected but not found should be clearly separated.

c410-f3r commented:
Now it is coding time!

Thank you very much for the review @petrochenkov! I will address everything and reach back as soon as possible

petrochenkov commented:
>Could you please separate out the change from `'{}'` to `` `{}` `` and the corresponding test changes, for separate review and FCP? That seems orthogonal to this change, and it'd reduce the amount that needs to be reviewed simultaneously.

\+1, the 'x' -> \`x\` changes are off-topic for this PR.

petrochenkov commented:
This also needs a second look from someone who is actually familiar with all this macro repetition logic, because I'm not.
cc @mark-i-m maybe

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,291 @@
+    crate fn parse(input: &tokenstream::TokenStream, sess: &ParseSess) -> Option<MetaVarExpr> {
+        let mut tts = input.trees();
+        match tts.next() {
+            Some(tokenstream::TokenTree::Token(token)) if token.is_ident() => {
```
Done

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,291 @@
+
+impl MetaVarExpr {
+    /// Attempt to parse a meta-variable expression from a token stream.
+    crate fn parse(input: &tokenstream::TokenStream, sess: &ParseSess) -> Option<MetaVarExpr> {
```
Converted from `Option` to `PResult`

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,291 @@
+                None
+            }
+            Some(tokenstream::TokenTree::Delimited(span, ..)) => {
+                let msg = "expected meta-variable expression, found delimited span";
```
Done

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,291 @@
+    Ident(Ident),
+
+    /// A literal argument.
+    Literal(token::Lit, Span),
```
Removed

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,291 @@
+
+//pub fn token_ident(token: &Token) -> (Ident, bool) {
+//    token.ident().expect(msg)
+//}
```
Ooopss

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,291 @@
+}
+
+/// Attempt to parse meta-variable expression arguments.
+fn parse_args(
```
Done

petrochenkov commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,291 @@
+    /// Parse a meta-variable `count` expression: `count(ident[, depth])`
+    fn parse_count(
+        span: Span,
+        input: impl Iterator<Item = tokenstream::TokenTree>,
```
No, I mean `&[tokenstream::TokenTree]` (or whatever is the actual type substituted here is).

c410-f3r replied:
Oh, I see. Thanks

mark-i-m commented:
I'll try to take a look soon.

c410-f3r commented:
@petrochenkov 

>  * which of the PR implements ignore, I don't see where variables used inside ${...} are marked as used, but apparently the variables used inside ignore are indeed counted as used because there's no warning.

My understanding is that ignored tokens are indeed counted but don't produce any output (https://github.com/rust-lang/rust/pull/93545/files#diff-727806e582042b1e7f196258a14707cc155d7ba36ab925066d3f7b43e6097b6fR511).

> * what is the difference between length and count.

https://github.com/rust-lang/rust/pull/93545/files#diff-23d875c394e16567144626114846b378fc49bbb8feedf5871399860ac5538744 has an extensive set of comments that hopefully will help understanding  the difference between both.

> * The variables named inside meta-variable expressions are defined.
> * The indices used inside meta-variable expressions are in range, I don't think there should be any silent saturating or wrapping behavior.

Did a fair amount of rewriting to improve these errors

> * +1, the 'x' -> `x` changes are off-topic for this PR.

Removed

mark-i-m reviewed (commented):
@c410-f3r Nice work navigating this very tricky and confusing code!

I left a bunch of comments in various places. My main concerns:
- I think one of the changes to `macro_check` is incomplete. I left a comment below.
- I think `count_repetitions` needs more comments, as it is not obvious what it does or why at a glance. I also left comments/suggestions there.

@petrochenkov 

> which of the PR implements ignore, I don't see where variables used inside ${...} are marked as used, but apparently the variables used inside ignore are indeed counted as used because there's no warning.

If you mean linting for unused meta-variables, I don't think we have any. It might be a good lint to add...

EDIT: I mean it doesn't exist for macros in general, and we should add it as a separate feature in a different PR.

If you mean checking for the number of iterations (e.g. so that `$(${ignore(x)} a)*` outputs `a` the number of times `$x` matched), that is done by `lockstep_iter_size`. The call to `lockstep_iter_size` from `transcribe` gets the number of repetitions matched by all variables repeating at that depth (and checks that all variables repeat the same amount).

c410-f3r commented on compiler/rustc_expand/src/mbe/macro_parser.rs:
```diff
@@ -392,7 +393,7 @@ fn nameize<I: Iterator<Item = NamedMatch>>(
                 Occupied(..) => return Err((sp, format!("duplicated bind name: {}", bind_name))),
             },
-            TokenTree::MetaVar(..) | TokenTree::Token(..) => (),
+            TokenTree::MetaVar(..) | TokenTree::MetaVarExpr(..) | TokenTree::Token(..) => (),
```
Well, it is up to you to decide if `MetaVarExpr` will panic :) 

Like actually panicking (`panic!( ... )`) or `span_diagnostic.span_err( ... );` ? In any case, I will do what is most appropriate.

However, dealing with `MetaVar` is kind of orthogonal. Maybe tackle everything in a following PR?

mark-i-m replied:
Leave it as-is for now. Maybe add a FIXME or open an issue to keep track of it?

c410-f3r replied:
Created a FIXME that will be fixed along side https://github.com/rust-lang/rust/issues/93902. Do you mind doing another round of review in a subsequent PR?

c410-f3r commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -352,6 +357,63 @@ impl LockstepIterSize {
+) -> PResult<'a, usize> {
+    fn count(matched: &NamedMatch, depth_opt: Option<usize>) -> usize {
+        match matched {
+            MatchedNonterminal(_) => 1,
```
Can I try it along side https://github.com/rust-lang/rust/issues/93902? Not sure if it is necessary to introduce an auxiliary counter to `fn count` in order to evaluate the maximum number of nested repetitions.

mark-i-m replied:
I think it should work as written, right? The max number of repetitions is depth_opt.

c410-f3r replied:
`depth_opt` is the user provided input index of `count($some_ident, some_user_input)`, which differs from the actual declared number of nested repetitions. For example, given `$( { $( $foo:ident )* } )*` and ``$( $( count(foo, 99999) )* )*, `depth_opt` would be 99999 while the actual number of "nested repetitions" is only 2

c410-f3r replied:
Did the out-of-bound check inside `count` as you suggested -> https://github.com/rust-lang/rust/pull/93545/files#diff-727806e582042b1e7f196258a14707cc155d7ba36ab925066d3f7b43e6097b6fR380-R407

mark-i-m replied:
Looks good!

c410-f3r commented on compiler/rustc_expand/src/mbe/quoted.rs:
```diff
@@ -118,6 +118,14 @@ pub(super) fn parse(
+/// Asks for the `macro_metavar_expr` feature if it is not already declared
+fn maybe_emit_macro_metavar_expr_feature(features: &Features, sess: &ParseSess, span: Span) {
+    if !features.macro_metavar_expr {
+        let msg = "Meta-variable expressions are unstable";
```
Renamed to lower-case

c410-f3r commented on src/test/ui/rfc-3086-metavar-expr/syntax-errors.rs:
```diff
@@ -0,0 +1,61 @@
+}
+
+macro_rules! curly__rhs_dollar__round {
+    ( $( $i:ident ),* ) => { ${ count($i) } };
```
Created two additional notes to help users -> https://github.com/rust-lang/rust/pull/93545/files#diff-be190d089aee3bc3b6fe37f11bb421bc9135101a611a9ed5e084f550334ab4b0R133-R155

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,159 @@
+    Ignore(Ident),
+
+    /// The index of the repetition at a particular depth, where 0 is the inner-most
+    /// repetition.
```
Done

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,159 @@
+    Index(usize),
+
+    /// The length of the repetition at a particular depth, where 0 is the inner-most
+    /// repetition.
```
Done

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,159 @@
+                    "length" => MetaVarExpr::Length(parse_depth(&mut iter, sess, ident.span)?),
+                    _ => {
+                        let msg = "unrecognised meta-variable expression";
+                        return Err(sess.span_diagnostic.struct_span_err(ident.span, msg));
```
Done

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,159 @@
+                ));
+            }
+            Some(tokenstream::TokenTree::Delimited(span, ..)) => {
+                let msg = "expected meta-variable expression, found delimited token group";
```
Done

c410-f3r commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,159 @@
+
+/// Eats the next comma in the token stream. If there is one, returns `true`.
+fn eat_comma(iter: &mut tokenstream::Cursor) -> bool {
+    if let Some(tt) = iter.next() {
```
Done

c410-f3r commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -385,6 +447,92 @@ fn lockstep_iter_size(
             }
         }
+        TokenTree::MetaVarExpr(_, ref expr) => {
+            expr.ident().into_iter().fold(LockstepIterSize::Unconstrained, |size, name| {
```
Used `let ... else ...` instead. Let me know if it is OK

c410-f3r commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -385,6 +447,92 @@ fn lockstep_iter_size(
+    span: Span,
+    ty: &str,
+) -> DiagnosticBuilder<'a> {
+    cx.struct_span_err(span, &format!("{ty} depth ({depth}) can not be equal or less than {max}"))
```
Done

c410-f3r commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -352,6 +357,63 @@ impl LockstepIterSize {
+                depth,
+                num_of_nested_repetitions,
+                sp.entire(),
+                "Count",
```
Done

c410-f3r commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -352,6 +357,63 @@ impl LockstepIterSize {
     }
 }
 
+fn count_repetitions<'a>(
```
Done

c410-f3r commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -352,6 +357,63 @@ impl LockstepIterSize {
+                    }
+                } else {
+                    ads.iter().map(|ad| count(ad, None)).sum()
+                }
```
Done

c410-f3r commented:
Thank you very much for the review, @mark-i-m! I really should have put more comments :)

c410-f3r commented:
@rustbot label S-waiting-on-review

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5183808051?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/02be3bea2379bcfe6ac3b76ec44f0c51102eb7de/checks/5183808051/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
..................................iii............................................................... 12600/12639
.......................................
failures:

---- [ui] ui/rfc-3086-metavar-expr/out-of-bounds-arguments.rs stdout ----

5    |              ^^^^^^^^^^^^^^^^
6 
6 
7 error: index depth can not be less than or equal to 3
-   --> $DIR/out-of-bounds-arguments.rs:20:18
9    |
9    |
10 LL |                 ${index(10)},
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu

12 
12 
13 error: length depth can not be less than or equal to 2
-   --> $DIR/out-of-bounds-arguments.rs:33:18
15    |
15    |
16 LL |                 ${length(10)}


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-3086-metavar-expr/out-of-bounds-arguments/out-of-bounds-arguments.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args rfc-3086-metavar-expr/out-of-bounds-arguments.rs`
error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/rfc-3086-metavar-expr/out-of-bounds-arguments.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-3086-metavar-expr/out-of-bounds-arguments" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-3086-metavar-expr/out-of-bounds-arguments/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: count depth can not be less than or equal to 4
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/out-of-bounds-arguments.rs:7:14
   |
LL |             ${count(foo, 10)},


error: index depth can not be less than or equal to 3
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/out-of-bounds-arguments.rs:19:18
   |
LL |                 ${index(10)},


error: length depth can not be less than or equal to 2
  --> /checkout/src/test/ui/rfc-3086-metavar-expr/out-of-bounds-arguments.rs:32:18
   |
LL |                 ${length(10)}

error: aborting due to 3 previous errors



------------------------------------------



failures:
    [ui] ui/rfc-3086-metavar-expr/out-of-bounds-arguments.rs
test result: FAILED. 12510 passed; 1 failed; 128 ignored; 0 measured; 0 filtered out; finished in 97.48s

Build completed unsuccessfully in 0:10:15
```

</details>

mark-i-m commented on compiler/rustc_ast/src/tokenstream.rs:
```diff
@@ -611,7 +611,7 @@ impl Cursor {
 
     pub fn look_ahead(&self, n: usize) -> Option<&TokenTree> {
-        self.stream.0[self.index..].get(n).map(|(tree, _)| tree)
+        self.stream.0.get(self.index..).and_then(|elem| elem.get(n)).map(|(tree, _)| tree)
```
why?

c410-f3r replied:
Hehehehe, couldn't resist. Sorry :P

Usually, I tend to avoid things that will panic at runtime but this indexing is perfectly safe. Let me remove this modification.

mark-i-m replied:
Thanks! Normally, I would agree, but in the compiler I would be wary of such changes. Often, the index/length of arrays means something. If they are not what we expect, weird stuff will happen that could be hard to debug. This seems to be especially true in the type system, too. Removing the panic would actually delay the error to happen somewhere else, where it could be more confusing.

mark-i-m commented on compiler/rustc_expand/src/mbe/macro_check.rs:
```diff
@@ -335,6 +336,7 @@ fn check_occurrences(
             let name = MacroRulesNormalizedIdent::new(name);
             check_ops_is_prefix(sess, node_id, macros, binders, ops, span, name);
         }
+        TokenTree::MetaVarExpr(..) => {}
```
@c410-f3r Yes, I believe `check_ops_is_prefix` would need to be modified to for meta-var-exprs. In particular, this function checks that meta-variables are defined when used, and that if they are used in (possibly nested) sequences, the Kleene operators and depth with which they use match those they are declared with. This would need to be modified slightly, I think as the meta-var-exprs can be used at any depth, iiuc...

Tracking this in another issue is fine, but it would be nice to fix this before the feature is stabilized.

c410-f3r replied:
Thank you for the information, @mark-i-m! If possible, I will try to fix this situation as well as the other remaining features in a following PR as soon as possible.

mark-i-m commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -385,6 +460,93 @@ fn lockstep_iter_size(
+    span: Span,
+    ty: &str,
+) -> DiagnosticBuilder<'a> {
+    cx.struct_span_err(span, &format!("{ty} depth can not be less than or equal to {max}"))
```
```suggestion
    cx.struct_span_err(span, &format!("{ty} depth must be less than or equal to {max}"))
```

or am I misunderstanding?

c410-f3r replied:
Wow... It was totally wrong. Thanks!

mark-i-m commented on src/test/ui/rfc-3086-metavar-expr/syntax-errors.stderr:
```diff
@@ -0,0 +1,135 @@
+note: Try removing `$`
+  --> $DIR/syntax-errors.rs:16:39
+   |
+LL |     ( $( $i:ident ),* ) => { ${ count($i) } };
+   |                                       ^
+
+note: meta-variable expressions do not expect arbitrary prefixes
+  --> $DIR/syntax-errors.rs:16:33
+   |
+LL |     ( $( $i:ident ),* ) => { ${ count($i) } };
+   |                                 ^^^^^
+
+error: a `ident` element was expected
+  --> $DIR/syntax-errors.rs:16:33
+   |
+LL |     ( $( $i:ident ),* ) => { ${ count($i) } };
+   |                                 ^^^^^
```
I think you want to attach the note to the error so that it appears after the error in the output.

mark-i-m replied:
But this note is exactly what I was thinking, and it is very helpful!

c410-f3r replied:
> I think you want to attach the note to the error so that it appears after the error in the output.

Hope it is good now

petrochenkov commented:
r? @mark-i-m 

(You can assign to me again when mark-i-m is happy with the result.)

c410-f3r commented:
@petrochenkov Thanks! I will get in touch afterwards 

@mark-i-m I tried to address the comments with the best of my capability. Let me know if there's anything else missing.

mark-i-m commented:
Sorry for the delay! I'll take a look soon.

mark-i-m commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,167 @@
+    Err(sess.span_diagnostic.struct_span_err(
+        span,
+        "could not find an expected `ident` element. \
+    Probably some number or special character was declared so try removing any arbitrary prefixes",
```
This should be in a note or something, I think.

mark-i-m commented on src/test/ui/rfc-3086-metavar-expr/syntax-errors.stderr:
```diff
@@ -0,0 +1,129 @@
+note: try removing `$`
+  --> $DIR/syntax-errors.rs:17:39
+   |
+LL |     ( $( $i:ident ),* ) => { ${ count($i) } };
+   |                                       ^
+
+error: could not find an expected `ident` element. Probably some number or special character was declared so try removing any arbitrary prefixes
+  --> $DIR/syntax-errors.rs:17:33
+   |
+LL |     ( $( $i:ident ),* ) => { ${ count($i) } };
+   |                                 ^^^^^
+
+error: expected one of: `*`, `+`, or `?`
+  --> $DIR/syntax-errors.rs:17:31
+   |
+LL |     ( $( $i:ident ),* ) => { ${ count($i) } };
+   |                               ^^^^^^^^^^^^^
+
```
I appreciate the time you've put into improving the errors, but I feel like these error messages still need work. @estebank or @petrochenkov are better suited to give you pointers on how to do it right.

1. The note should be attached to a particular error. I believe your use of `span_note_without_error` is the problem here.
2. I don't find the second error very helpful. If I received that error as a user, I wouldn't really know what to do. Also, we typically don't put suggestions in the error message itself.
3. ```expected one of: `*`, `+`, or `?` ``` This seems wrong. I feel like the error message I would expect is something like:

```
error: expected ident found `$`
  --> $DIR/syntax-errors.rs:17:31
   |
LL |     ( $( $i:ident ),* ) => { ${ count($i) } };
   |                                       ^
```

c410-f3r replied:
> expected one of: `*`, `+`, or `?` This seems wrong. I feel like the error message I would expect is something like:

`parse_tree` now returns early to avoid unrelated meta-variables expression errors -> https://github.com/rust-lang/rust/pull/93545/files#diff-152c18de02d7a7e6ad577226b23ae10329c206af65287b78f35cc7f37e7ba699R187-R189

> The note should be attached to a particular error. I believe your use of span_note_without_error is the problem here.

> I don't find the second error very helpful. If I received that error as a user, I wouldn't really know what to do. Also, we typically don't put suggestions in the error message itself.

![Capture d’écran de 2022-02-22 18-40-00](https://user-images.githubusercontent.com/17877264/155224322-0a819da2-2583-47ec-8d63-db2179038ed6.png)

petrochenkov replied:
I'm fine with the errors as they are.

mark-i-m reviewed (approved):
The core logic looks good to me!

I left some comments about minor concerns below, most notably the error messages. I think this PR can be merged, but those things should be fixed before the feature is stabilized.

@petrochenkov feel free to r=me

r? @petrochenkov

mark-i-m commented:
Right... forgot bors doesn't look in reviews...

r? @petrochenkov

c410-f3r commented on src/tools/tidy/src/ui_tests.rs:
```diff
@@ -7,7 +7,7 @@ use std::path::Path;
 const ENTRY_LIMIT: usize = 1000;
 // FIXME: The following limits should be reduced eventually.
-const ROOT_ENTRY_LIMIT: usize = 982;
+const ROOT_ENTRY_LIMIT: usize = 983;
```
Because a new folder was added to accommodate the tests for this feature and `src/test/ui` currently can not exceed more than 982 entries.

This change is funny because I was trimming down the whole test suite -> https://github.com/rust-lang/rust/issues/73494

@petrochenkov Can we please remove `ROOT_ENTRY_LIMIT` to keep it in agreement with the remaining folders (1000 entries max)? Full separation of standalone unit test files into appropriated directories can be performed afterwards and does not depend on `ROOT_ENTRY_LIMIT`

petrochenkov replied:
I would prefer to move the whole `rfc-3086-metavar-expr` subdirectory to `ui/macros`.

c410-f3r replied:
Moved to `ui/macros` but I still insist on removing `ROOT_ENTRY_LIMIT` in another PR because people will simply keep increasing the value and it probably won't take long to exceed 1000 entries (again). Besides, its existence was due to its large amount of files which is not the case anymore.

`ui/parser` is also nearing 1000 entries so IMO, all directories should be constantly monitored and fragmented over time without any special handling.

c410-f3r commented on compiler/rustc_expand/src/mbe/transcribe.rs:
```diff
@@ -385,6 +468,93 @@ fn lockstep_iter_size(
+            let default_rslt = LockstepIterSize::Unconstrained;
+            let Some(ident) = expr.ident() else { return default_rslt; };
+            let name = MacroRulesNormalizedIdent::new(ident.clone());
+            match lookup_cur_matched(name, interpolations, repeats) {
+                Some(matched) => match matched {
+                    MatchedNonterminal(_) => default_rslt,
+                    MatchedSeq(ref ads) => {
+                        default_rslt.with(LockstepIterSize::Constraint(ads.len(), name))
+                    }
+                },
+                _ => default_rslt,
+            }
```
Changed

c410-f3r commented:
@mark-i-m Thank you very much for the review and if desired, feel free to suggest more changes

petrochenkov commented:
I've noticed one unaddressed comment here - https://github.com/rust-lang/rust/pull/93545#discussion_r800191457.

c410-f3r commented:
> I've noticed one unaddressed comment here - [#93545 (comment)](https://github.com/rust-lang/rust/pull/93545#discussion_r800191457).

Based on https://github.com/rust-lang/rust/pull/93545#issuecomment-1049758957 and other comments, this case will be handled in a following PR

c410-f3r committed:
Implement macro meta-variable expressions

petrochenkov commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,180 @@
+impl MetaVarExpr {
+    /// Attempt to parse a meta-variable expression from a token stream.
+    crate fn parse<'sess>(
+        input: &tokenstream::TokenStream,
```
Nit: could you import names instead of using two-segment paths like `tokenstream::TokenStream` or `tokenstream::TokenTree`>?
(This may apply to other files too.)

petrochenkov commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,180 @@
+                    "length" => MetaVarExpr::Length(parse_depth(&mut iter, sess, ident.span)?),
+                    _ => {
+                        let msg = "unrecognised meta-variable expression. Supported expressions \
+                        are count, ignore, index and length";
```
The suggestion is better moved from the main message to `span_suggestion` attached to the error.

petrochenkov replied:
Also, no test case for this error.

petrochenkov commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,180 @@
+                    }
+                };
+                if let Some(arg) = iter.next() {
+                    let msg = "unexpected meta-variable expression argument";
```
No test case for this error.

petrochenkov commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,180 @@
+                return Err(sess.span_diagnostic.struct_span_err(
+                    token.span,
+                    &format!(
+                        "expected meta-variable expression, found `{}`",
```
No test case for this error.

petrochenkov commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,180 @@
+                    ),
+                ));
+            }
+            _ => return Err(sess.span_diagnostic.struct_err("expected meta-variable expression"))
```
No test case for this error.

petrochenkov commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,180 @@
+                let mut err = err_fn();
+                err.span_suggestion(
+                    token.span,
+                    &format!("Try removing `{}`", pprust::token_to_string(&token)),
```
```suggestion
                    &format!("try removing `{}`", pprust::token_to_string(&token)),
```

petrochenkov replied:
Also, this suggestion doesn't make sense in a general case, I'd report it only if the current token is `$` and the next token is an identifier (if that was the original goal).

petrochenkov commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,180 @@
+    span: Span,
+) -> PResult<'sess, Ident> {
+    let err_fn =
+        || sess.span_diagnostic.struct_span_err(span, "could not find an expected `ident` element");
```
This looks like a name resolution error or something.
Parser reports something like "expected identifier, found `...`" in this situation.

petrochenkov commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,180 @@
+                );
+                return Err(err);
+            }
+            tokenstream::TokenTree::Delimited(delim_span, _, _) => {
```
This error case looks highly useless and unlikely to happen in practice, I'd remove it.

petrochenkov commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,180 @@
+                err.span_suggestion(
+                    token.span,
+                    &format!("Try removing `{}`", pprust::token_to_string(&token)),
+                    <_>::default(),
```
```suggestion
                    String::new(),
```

petrochenkov commented on compiler/rustc_expand/src/mbe/metavar_expr.rs:
```diff
@@ -0,0 +1,180 @@
+    }) = tt else {
+        return Err(sess.span_diagnostic.struct_span_err(
+            span,
+            "meta-expression depth must be a literal"
```
No test case for this error.

petrochenkov replied:
```suggestion
            "meta-variable expression depth must be a literal"
```

petrochenkov commented:
I started reviewing this, but didn't finish - the PR is too large.

Could you split it into 2 PRs:
- The first one introduces the meta-variable expression infra, and one of the expression kinds, e.g. `ignore` as the simplest one.
- The second one implements the remaining expression kinds.

c410-f3r commented on src/test/ui/macros/rfc-3086-metavar-expr/count-and-length-are-distinct.rs:
```diff
@@ -0,0 +1,271 @@
+    );
+
+    // It is possible to say, to some degree, that count is an "amalgamation" of length (see
+    // each length line result and compare them with the count results)
```
Heh, I really tried my best on these examples :)

Maybe it is easier to understand this way: `count` counts every element repetition of every nested loop while `length` only counts the repetition of a single loop.

For example, given 5 nested loops declared on the LHS and a macro call specifying 2 outermost repetitions, `${length(4)}` [index `4` because there are 5 nested loops (`length` starts in the innermost repetition)] will output `2` and `${count(ident, 0)}` [index `0` because `count` starts in the outermost (same as `${count(ident)}`)] will output the sum of all `ident` (whatever token) repetitions of all nested loops (could be 10, 2304 or 999999999, idk).

c410-f3r commented on src/test/ui/macros/rfc-3086-metavar-expr/requiring-feature.rs:
```diff
@@ -0,0 +1,15 @@
+}
+
+macro_rules! a {
+    ( $$a:ident ) => {
```
`SS` in the LHS  will allow repetitions within nested loops without ugly hacks

```rust
macro_rules! nested {
    ( $a:ident ) => {
        macro_rules! $a {
            ( $$( $b:ident ),* ) => { // <- Here
                $$( ... )*
            };
        }
    };
}
```

The RFC also makes explicit reference of them in the LHS -> https://github.com/markbt/rfcs/blob/macro_metavar_expr/text/0000-macro-metavar-expr.md#dollar-dollar

petrochenkov replied:
This is not LHS, it's RHS of `nested`.
When it becomes LHS (of `$a`) `$$` is already turned into a single `$`.

c410-f3r replied:
Hum... I see now. Thanks

c410-f3r commented:
Addresses the comments here and created https://github.com/rust-lang/rust/pull/94368 as requested
